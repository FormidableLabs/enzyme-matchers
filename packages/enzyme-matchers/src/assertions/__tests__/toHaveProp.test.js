const PropTypes = require('prop-types');

const toHaveProp = require('../toHaveProp');

function User(props) {
  return (
    <div>
      {props.name}
      {props.email}
      {props.arrayProp}
      {props.objectProp.foo}
      {props.falsy}
    </div>
  );
}

User.propTypes = {
  name: PropTypes.string,
  email: PropTypes.string,
  arrayProp: PropTypes.array,
  objectProp: PropTypes.object,
  falsy: PropTypes.bool,
};

const fn = () => {};

function Fixture() {
  return (
    <User
      name="blaine"
      email="blainekasten@gmail.com"
      arrayProp={[1, 2, 3]}
      objectProp={{ foo: 'bar' }}
      fn={fn}
      falsy={false}
    />
  );
}

describe('toHaveProp', () => {
  [shallow, mount].forEach(builder => {
    describe(builder.name, () => {
      function build() {
        const wrapper = builder(<Fixture />);
        const user = wrapper.find(User);
        const truthyResults = toHaveProp(user, 'arrayProp', [1, 2, 3]);
        const falsyResults = toHaveProp(user, 'arrayProp', [4, 5, 6]);
        const anyValueResults = toHaveProp(wrapper, 'arrayProp');

        return {
          truthyResults,
          falsyResults,
          anyValueResults,
        };
      }

      it('returns the pass flag properly', () => {
        const { truthyResults, falsyResults } = build();

        expect(truthyResults.pass).toBe(true);
        expect(falsyResults.pass).toBe(false);
      });

      it(`returns the message with the proper pass verbage (${builder.name})`, () => {
        const { truthyResults } = build();
        expect(truthyResults.message).toMatchSnapshot();
      });

      it(`returns the message with the proper fail verbage (${builder.name})`, () => {
        const { truthyResults } = build();
        expect(truthyResults.negatedMessage).toMatchSnapshot();
      });

      it(`provides contextual information for the message (${builder.name})`, () => {
        const { truthyResults } = build();
        expect(truthyResults.contextualInformation).toMatchSnapshot();
      });

      it(`returns the "any value" message with the proper pass verbage (${builder.name})`, () => {
        const { anyValueResults } = build();
        expect(anyValueResults.message).toMatchSnapshot();
      });

      it(`returns the "any value" message with the proper fail verbage (${builder.name})`, () => {
        const { anyValueResults } = build();
        expect(anyValueResults.negatedMessage).toMatchSnapshot();
      });

      it(`provides "any value" contextual information for the message (${builder.name})`, () => {
        const { anyValueResults } = build();
        expect(anyValueResults.contextualInformation).toMatchSnapshot();
      });

      it('can validate arrays', () => {
        const wrapper = builder(<Fixture />);
        const user = wrapper.find(User);
        const truthy = toHaveProp(user, 'arrayProp', [1, 2, 3]);
        const falsy = toHaveProp(user, 'arrayProp', [4, 5, 6]);

        expect(truthy.pass).toBe(true);
        expect(falsy.pass).toBe(false);
      });

      it('can validate objects', () => {
        const wrapper = builder(<Fixture />);
        const user = wrapper.find(User);
        const truthy = toHaveProp(user, 'objectProp', { foo: 'bar' });
        const falsy = toHaveProp(user, 'objectProp', { foo: 'BOO' });

        expect(truthy.pass).toBe(true);
        expect(falsy.pass).toBe(false);
      });

      it('works with falsy props', () => {
        const wrapper = builder(<Fixture />);
        const { pass } = toHaveProp(wrapper.find(User), 'falsy', false);

        expect(pass).toBe(true);
      });

      it('works with functions', () => {
        const wrapper = builder(<Fixture />);
        const { pass } = toHaveProp(wrapper.find(User), 'fn', fn);
        const { pass: fail } = toHaveProp(wrapper.find(User), 'fn', () => {});

        expect(pass).toBe(true);
        expect(fail).toBe(false);
      });

      it('works without a prop value', () => {
        const wrapper = builder(<Fixture />);
        const truthy = toHaveProp(wrapper.find(User), 'name');

        expect(truthy.pass).toBe(true);
      });

      it('works with undefined value', () => {
        const wrapper = builder(<Fixture />);
        const falsy = toHaveProp(wrapper.find(User), 'name', undefined);

        expect(falsy.pass).toBe(false);
      });
    });
  });
});
