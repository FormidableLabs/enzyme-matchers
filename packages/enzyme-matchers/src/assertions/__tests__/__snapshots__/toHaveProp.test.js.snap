// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveProp mount provides "any value" contextual information for the message (mount) 1`] = `
Object {
  "actual": "Actual props: [33marrayProp[39m[90m:[39m [33m[31mundefined[33m[39m",
}
`;

exports[`toHaveProp mount provides contextual information for the message (mount) 1`] = `
Object {
  "actual": "Actual props: [33marrayProp[39m[90m:[39m [33m[90m[90m[[90m
  [34m1[90m
  [34m2[90m
  [34m3[90m
[90m][90m[33m[39m",
  "expected": "Expected props: [33marrayProp[39m[90m:[39m [33m[90m[90m[[90m
  [34m1[90m
  [34m2[90m
  [34m3[90m
[90m][90m[33m[39m",
}
`;

exports[`toHaveProp mount returns the "any value" message with the proper fail verbage (mount) 1`] = `"Expected <Fixture> not to receive prop \\"arrayProp\\", but it did."`;

exports[`toHaveProp mount returns the "any value" message with the proper pass verbage (mount) 1`] = `"Expected <Fixture> to receive prop \\"arrayProp\\", but it did not."`;

exports[`toHaveProp mount returns the message with the proper fail verbage (mount) 1`] = `"Expected <User> to not match for props \\"\\", but it did."`;

exports[`toHaveProp mount returns the message with the proper pass verbage (mount) 1`] = `"Expected <User> to match for props \\"\\", but it did not."`;

exports[`toHaveProp shallow provides "any value" contextual information for the message (shallow) 1`] = `
Object {
  "actual": "Actual props: [33marrayProp[39m[90m:[39m [33m[90m[90m[[90m
  [34m1[90m
  [34m2[90m
  [34m3[90m
[90m][90m[33m[39m",
}
`;

exports[`toHaveProp shallow provides contextual information for the message (shallow) 1`] = `
Object {
  "actual": "Actual props: [33marrayProp[39m[90m:[39m [33m[90m[90m[[90m
  [34m1[90m
  [34m2[90m
  [34m3[90m
[90m][90m[33m[39m",
  "expected": "Expected props: [33marrayProp[39m[90m:[39m [33m[90m[90m[[90m
  [34m1[90m
  [34m2[90m
  [34m3[90m
[90m][90m[33m[39m",
}
`;

exports[`toHaveProp shallow returns the "any value" message with the proper fail verbage (shallow) 1`] = `"Expected <User> not to receive prop \\"arrayProp\\", but it did."`;

exports[`toHaveProp shallow returns the "any value" message with the proper pass verbage (shallow) 1`] = `"Expected <User> to receive prop \\"arrayProp\\", but it did not."`;

exports[`toHaveProp shallow returns the message with the proper fail verbage (shallow) 1`] = `"Expected <User> to not match for props \\"\\", but it did."`;

exports[`toHaveProp shallow returns the message with the proper pass verbage (shallow) 1`] = `"Expected <User> to match for props \\"\\", but it did not."`;
